### ISSUE TITLE
[BUG] System.Net.ProtocolViolationException while posting message to queue

### ISSUE THREAD

**Issue by fokkog** [Azure/azure-sdk-for-net#16111]
**Describe the bug**
While posting a message containing a special character to a queue, we receive "Unhandled Exception: System.Net.ProtocolViolationException: Bytes to be written to the stream exceed the Content-Length bytes size specified".

This seems a variation on [this bug](https://github.com/Azure/azure-sdk-for-net/issues/11358). Our code is running on .NET Framework instead of .NET Core. This is a completely unexpected regression from v9.x SDK (WindowsAzure.Storage package) / v11.x SDK (Microsoft.Azure.Storage.Queue package).

**Expected behavior**
Posting a message does not fail randomly due to message content.

**Actual behavior (include Exception or Stack Trace)**
See description above, full stack trace is:
```
Unhandled Exception: System.Net.ProtocolViolationException: Bytes to be written to the stream exceed the Content-Length bytes size specified.
   at System.Net.ConnectStream.InternalWrite(Boolean async, Byte[] buffer, Int32 offset, Int32 size, AsyncCallback callback, Object state)
   at System.Net.ConnectStream.BeginWrite(Byte[] buffer, Int32 offset, Int32 size, AsyncCallback callback, Object state)
   at System.IO.Stream.<>c.<BeginEndWriteAsync>b__53_0(Stream stream, ReadWriteParameters args, AsyncCallback callback, Object state)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncTrim[TInstance,TArgs](TInstance thisRef, TArgs args, Func`5 beginMethod, Func`3 endMethod)
   at System.IO.Stream.BeginEndWriteAsync(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.Stream.WriteAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at Azure.Core.RequestContent.ArrayContent.<WriteToAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpClientTransport.PipelineRequest.PipelineContentAdapter.<SerializeToStreamAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpClientTransport.<ProcessAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpClientTransport.Process(HttpMessage message)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.RequestActivityPolicy.<ProcessNextAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.RequestActivityPolicy.<ProcessAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.RequestActivityPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.ResponseBodyPolicy.<ProcessAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.ResponseBodyPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.LoggingPolicy.<ProcessAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.LoggingPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.RetryPolicy.<ProcessAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Azure.Core.Pipeline.RetryPolicy.<ProcessAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.RetryPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipeline.Send(HttpMessage message, CancellationToken cancellationToken)
   at Azure.Storage.Queues.QueueRestClient.Messages.<EnqueueAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Queues.QueueClient.<SendMessageInternal>d__75.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Queues.QueueClient.SendMessage(String messageText, Nullable`1 visibilityTimeout, Nullable`1 timeToLive, CancellationToken cancellationToken)
   at Azure.Storage.Queues.QueueClient.SendMessage(String messageText)
...
```

**To Reproduce**
Instantiate a queue client and invoke:
`client.SendMessage("{\"name\":\"Müller\"}");`

**Environment:**
 - Name and version of the Library package used: Azure.Storage.Queues 12.4.2
 - Hosting platform or OS and .NET runtime version: Windows Server 2019 / .NET Framework 4.7.2
 - IDE and version : not relevant

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by furlongmt** [comment]
Hi,

I'm also facing the same issue after updating from V11->V12. I've noticed that the message is successfully added to the queue even when the exception is thrown.

Thanks,
Matt

**Comment by fokkog** [comment]
After rechecking: in our case no message can be found on the queue when this exception occurs.

I am curious what the MS position is on this bug, 3 weeks without response and counting. To compare, earlier we reported [another regression](https://github.com/Azure/azure-sdk-for-net/issues/15628) from the v11.x to v12.x upgrade and that was addressed instantly.

**Comment by amnguye** [comment]
Related issue: https://github.com/Azure/azure-sdk-for-net/issues/10312

Interesting I would have thought that the [PR](https://github.com/Azure/azure-sdk-for-net/pull/16689) to introduce BinaryData would have resolved this issue because it would BaseEncode64 the message and prevent the "Bytes to be written to the stream exceed the Content-Length bytes size specified" error from happening.

Let me see if I can reproduce the issue, and find a mitigation with the BinaryData solution to help.

**Comment by kasobol-msft** [comment]
That happens because V11 and below implicitly Base64-encoded messages while V12 doesn't. The default has changed.
While migrating from V11 to V12 https://docs.microsoft.com/en-us/dotnet/api/azure.storage.queues.queueclientoptions.messageencoding?view=azure-dotnet should be used to align behavior between V11 and V12. I.e. if you were using V11 on defaults you should use `QueueMessageEncoding.Base64` in `QueueClientOptions ` while creating `QueueClient`.

**Comment by fokkog** [comment]
Thanks for the suggested workaround. We had reached the same conclusion, and resorted to base64 encoding before posting the message and base64 decoding after retrieving the message.

IMO this new option should be the default for everyone, not just users who upgrade from v11. Without this, there are unclear and undocumented constraints around the kind of messageText that can be passed into [SendMessage](https://docs.microsoft.com/en-us/dotnet/api/azure.storage.queues.queueclient.sendmessage?view=azure-dotnet#Azure_Storage_Queues_QueueClient_SendMessage_System_String_). We were unlucky enough to not discover this regression during testing. Instead we only found this in production, where roughly 1 in 100 message transfers failed due to this bug.
